docs/config-local-settings.rst
Graphite-web's local_settings.py
================================
Graphite-web uses the convention of importing a ``local_settings.py`` file from the webapp
``settings.py`` module. This is where Graphite-web's runtime configuration is loaded from.

Config File Location
--------------------
``local_settings.py`` is generally located within the main ``graphite`` module where the
webapp's code lives. In the :ref:`default installation layout <default-installation-layout>` this
is ``/opt/graphite/webapp/graphite/local_settings.py``. Alternative locations can be used by
symlinking to this path or by ensuring the module can be found within the Python module search
path.

General Settings
----------------
TIME_ZONE
  `Default: America/Chicago`

  Set your local timezone. Timezone is specifed using
  `zoneinfo names <http://en.wikipedia.org/wiki/Zoneinfo#Names_of_time_zones>`_.

DOCUMENTATION_URL
  `Default: http://graphite.readthedocs.org/`

  Overrides the `Documentation` link used in the header of the Graphite Composer

LOG_RENDERING_PERFORMANCE
  `Default: False`

  Triggers the creation of ``rendering.log`` which logs timings for calls to the :doc:`render_api`

LOG_CACHE_PERFORMANCE
  `Default: False`

  Triggers the creation of ``cache.log`` which logs timings for remote calls to `carbon-cache` as
  well as Request Cache (memcached) hits and misses.

LOG_METRIC_ACCESS
  `Default: False`

  Trigges the creation of ``metricaccess.log`` which logs access to `Whisper` and `RRD` data files

DEBUG = True
  `Default: False`

  Enables generation of detailed Django error pages. See `Django's documentation <https://docs.djangoproject.com/en/dev/ref/settings/#debug>`_ for details

FLUSHRRDCACHED
  `Default: <unset>`

  If set, executes ``rrdtool flushcached`` before fetching data from RRD files. Set to the address
  or socket of the rrdcached daemon. Ex: ``unix:/var/run/rrdcached.sock``

MEMCACHE_HOSTS
  `Default: []`

  If set, enables the caching of calculated targets (including applied functions) and rendered images. If running
  a cluster of Graphite webapps, each webapp should have the exact same values for this setting to prevent unneeded
  cache misses.

  Set this to the list of memcached hosts. Ex: ``['10.10.10.10:11211', '10.10.10.11:11211', '10.10.10.12:11211']``

DEFAULT_CACHE_DURATION
  `Default: 60`

  Default expiration of cached data and images.


Filesystem Paths
----------------
These settings configure the location of Graphite-web's additional configuration files, static content,
and data. These need to be adjusted if Graphite-web is installed outside of the
:ref:`default installation layout <default-installation-layout>`.

GRAPHITE_ROOT
  `Default: /opt/graphite`
  The base directory for the Graphite install. This setting is used to shift the Graphite install
  from the default base directory which keeping the :ref:`default layout <default-installation-layout>`.
  The paths derived from this setting can be individually overridden as well

CONF_DIR
  `Default: GRAPHITE_ROOT/conf`
  The location of additional Graphite-web configuration files

STORAGE_DIR
  `Default: GRAPHITE_ROOT/storage`
  The base directory from which WHISPER_DIR, RRD_DIR, LOG_DIR, and INDEX_FILE default paths
  are referenced

CONTENT_DIR
  `Default: See below`
  The location of Graphite-web's static content. This defaults to ``content/`` two parent
  directories up from ``settings.py``. In the :ref:`default layout <default-installation-layout>`
  this is ``/opt/graphite/webapp/content``

DASHBOARD_CONF
  `Default: CONF_DIR/dashboard.conf`
  The location of the Graphite-web Dashboard configuration

GRAPHTEMPLATES_CONF
  `Default: CONF_DIR/graphTemplates.conf`
  The location of the Graphite-web Graph Template configuration

WHISPER_DIR
  `Default: /opt/graphite/storage/whisper`
  The location of Whisper data files

RRD_DIR
  `Default: /opt/graphite/storage/rrd`
  The location of RRD data files

STANDARD_DIRS
  `Default: [WHISPER_DIR, RRD_DIR]`
  The list of directories searched for data files. By default, this is the value of WHISPER_DIR
  and RRD_DIR (if rrd support is detected). If this setting is defined, the WHISPER_DIR and RRD_DIR
  settings have no effect.

LOG_DIR
  `Default: STORAGE_DIR/log/webapp`
  The directory to write Graphite-web's log files. This directory must be writable by the user
  running the Graphite-web webapp

INDEX_FILE
  `Default: /opt/graphite/storage/index`
  The location of the search index file. This file is generated by the `build-index.sh` script and
  must be writable by the user running the Graphite-web webap

Email Configuration
-------------------
These settings configure Django's email functionality which is used for emailing rendered graphs.
See the `Django documentation <https://docs.djangoproject.com/en/dev/topics/email/>`_ for
further detail on these settings

EMAIL_BACKEND
  `Default: django.core.mail.backends.smtp.EmailBackend`
  Set to ``django.core.mail.backends.dummy.EmailBackend`` to drop emails on the floor and effectively
  disable email features.

EMAIL_HOST
  `Default: localhost`

EMAIL_PORT
  `Default: 25`

EMAIL_HOST_USER
  `Default: ''`

EMAIL_HOST_PASSWORD
  `Default: ''`

EMAIL_USE_TLS
  `Default: False`

Authentication Configuration
----------------------------
These settings insert additional backends to the
`AUTHENTICATION_BACKENDS <https://docs.djangoproject.com/en/dev/ref/settings/#authentication-backends>`_
and `MIDDLEWARE_CLASSES <https://docs.djangoproject.com/en/dev/ref/settings/#std:setting-MIDDLEWARE_CLASSES>`_ settings. Additional authentication schemes are possible by manipulating these lists directly.

LDAP
^^^^
These settings configure a custom LDAP authentication backend provided by Graphite. Additional
settings to the ones below are configurable setting the LDAP module's global options using
``ldap.set_option``. See the `module documentation <http://python-ldap.org/>`_ for more details.

.. code-block:: none

  # SSL Example
  import ldap
  ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_ALLOW)
  ldap.set_option(ldap.OPT_X_TLS_CACERTDIR, "/etc/ssl/ca")
  ldap.set_option(ldap.OPT_X_TLS_CERTFILE, "/etc/ssl/mycert.pem")
  ldap.set_option(ldap.OPT_X_TLS_KEYFILE, "/etc/ssl/mykey.pem")

USE_LDAP_AUTH
  `Default: False`

LDAP_SERVER
  `Default: ''`

  Set the LDAP server here or alternativly in ``LDAP_URI``

LDAP_PORT
  `Default: 389`

  Set the LDAP server port here or alternativly in ``LDAP_URI``

LDAP_URI
  `Default: None`

  Sets the LDAP server URI. E.g. ``ldaps://ldap.mycompany.com:636``

LDAP_SEARCH_BASE
  `Default: ''`

  Sets the LDAP search base. E.g. ``OU=users,DC=mycompany,DC=com``

LDAP_BASE_USER
  `Default: ''`

  Sets the base LDAP user to bind to the server with. E.g.
  ``CN=some_readonly_account,DC=mycompany,DC=com``

LDAP_BASE_PASS
  `Default: ''`

  Sets the password of the base LDAP user to bind to the server with.

LDAP_USER_QUERY
  `Default: ''`

  Sets the LDAP query to return a user object where ``%s`` substituted with the
  user id. E.g. ``(username=%s)`` or ``(sAMAccountName=%s)`` (Active Directory)

Other Authentications
^^^^^^^^^^^^^^^^^^^^^
REMOTE_USER_AUTHENTICATION
  `Default: False`

  Enables the use of the Django `RemoteUserBackend` authentication backend. See the
  `Django documentation <https://docs.djangoproject.com/en/dev/howto/auth-remote-user/>`_ for
  further details

LOGIN_URL
  `Default: /account/login`

  Modifies the URL linked in the `Login` link in the Composer interface. This is useful
  for directing users to an external authentication link such as for Remote User authentication
  or a backend such as `django_openid_auth <https://launchpad.net/django-openid-auth>`_

Dashboard Authorization Configuration
-------------------------------------
These settings control who is allowed to save and delete dashboards.  By default anyone
can perform these actions, but by setting DASHBOARD_REQUIRE_AUTHENTICATION, users must at
least be logged in to do so.  The other two settings allow further restriction of who is
able to perform these actions. Users who are not suitably authorized will still be able to
use and change dashboards, but will not be able to save changes or delete dashboards.

DASHBOARD_REQUIRE_AUTHENTICATION
  `Default: False`

  If set to True, dashboards can only be saved and deleted by logged in users.

DASHBOARD_REQUIRE_EDIT_GROUP
  `Default: None`

  If set to the name of a user group, dashboards can only be saved and deleted by logged-in users
  who are members of this group.  Groups can be set in the Django Admin app, or in LDAP.

  Note that DASHBOARD_REQUIRE_AUTHENTICATION must be set to true - if not, this setting is ignored.

DASHBOARD_REQUIRE_PERMISSIONS
  `Default: False`

  If set to True, dashboards can only be saved or deleted by users having the appropriate
  (change or delete) permission (as set in the Django Admin app). These permissions can be set at
  the user or group level.  Note that Django's 'add' permission is not used.
  
  Note that DASHBOARD_REQUIRE_AUTHENTICATION must be set to true - if not, this setting is ignored.

Database Configuration
----------------------
The following configures the Django database settings. Graphite uses the database for storing user
profiles, dashboards, and for the Events functionality. Graphite uses an Sqlite database file located
at ``STORAGE_DIR/graphite.db`` by default. If running multiple Graphite-web instances, a database
such as PostgreSQL or MySQL is required so that all instances may share the same data source.

.. note ::
  As of Django 1.2, the database configuration is specified by the DATABASES
  dictionary instead of the old ``DATABASE_*`` format. Users must use the new
  specification to have a working database.

See the
`Django documentation <https://docs.djangoproject.com/en/dev/ref/settings/#databases>`_
for full documentation of the DATABASE setting.

.. note ::
   Remember, setting up a new database requires running ``manage.py syncdb`` to create the initial
   schema

Cluster Configuration
---------------------
These settings configure the Graphite webapp for clustered use. When ``CLUSTER_SERVERS`` is set,
metric browse and render requests will cause the webapp to query other webapps in CLUSTER_SERVERS for
matching metrics. Graphite will use only one successfully matching response to render data. This means
that metrics may only live on a single server in the cluster unless the data is consistent on both
sources (e.g. with shared SAN storage). Duplicate metric data existing in multiple locations will
*not* be combined.

CLUSTER_SERVERS
  `Default: []`

  The list of IP addresses and ports of remote Graphite webapps in a cluster. Each of these servers
  should have local access to metric data to serve. The first server to return a match for a query
  will be used to serve that data. Ex: ["10.0.2.2:80", "10.0.2.3:80"]

REMOTE_STORE_FETCH_TIMEOUT
  `Default: 6`

  Timeout for remote data fetches in seconds

REMOTE_STORE_FIND_TIMEOUT
  `Default: 2.5`

  Timeout for remote find requests (metric browsing) in seconds

REMOTE_STORE_RETRY_DELAY
  `Default: 60`

  Time in seconds to blacklist a webapp after a timed-out request

REMOTE_FIND_CACHE_DURATION
  `Default: 300`

  Time to cache remote metric find results in seconds

REMOTE_RENDERING
  `Default: False`

  Enable remote rendering of images and data (JSON, et al.) on remote Graphite webapps. If this
  is enabled, ``RENDERING_HOSTS`` must be configured below

RENDERING_HOSTS
  `Default: []`

  List of IP addresses and ports of remote Graphite webapps used to perform rendering. Each webapp
  must have access to the same data as the Graphite webapp which uses this setting either through
  shared local storage or via ``CLUSTER_SERVERS``. Ex: ["10.0.2.4:80", "10.0.2.5:80"]

REMOTE_RENDER_CONNECT_TIMEOUT
  `Default: 1.0`

  Connection timeout for remote rendering requests in seconds

CARBONLINK_HOSTS
  `Default: [127.0.0.1:7002]`

  If multiple carbon-caches are running on this machine, each should be listed here so that the Graphite
  webapp may query the caches for data that has not yet been persisted. Remote carbon-cache instances in a
  multi-host clustered setup should *not* be listed here. Instance names should be listed
  as applicable. Ex: ['127.0.0.1:7002:a','127.0.0.1:7102:b', '127.0.0.1:7202:c']

CARBONLINK_TIMEOUT
  `Default: 1.0`

  Timeout for carbon-cache cache queries in seconds

Additional Django Settings
--------------------------
The ``local_settings.py.example`` shipped with Graphite-web imports ``app_settings.py`` into the
namespace to allow further customization of Django. This allows the setting or customization of
standard `Django settings <https://docs.djangoproject.com/en/dev/ref/settings/>`_ and the installation
and configuration of additional `middleware <https://docs.djangoproject.com/en/dev/topics/http/middleware/>`_.
To manipulate these settings, ensure ``app_settings.py`` is imported as such:

.. code-block:: python

   from graphite.app_settings import *

The most common settings to manipulate are ``INSTALLED_APPS``, ``MIDDLEWARE_CLASSES``, and ``AUTHENTICATION_BACKENDS`` 


debian/changelog
yabs-graphite-ceres (0.2.0.26) unstable; urgency=low

  * removed useless files in opt

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 06 Dec 2013 13:57:22 +0400

yabs-graphite-ceres (0.2.0.25) precise; urgency=low

  * added construct button in dashboard

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 04 Dec 2013 23:38:20 +0400

yabs-graphite-ceres (0.2.0.24) unstable; urgency=low

  * moved to github
  * few minor fixes 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 04 Dec 2013 21:45:56 +0400

yabs-graphite-ceres (0.2.0.23) unstable; urgency=low

  * depends fix

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 21:11:16 +0400

yabs-graphite-ceres (0.2.0.22) unstable; urgency=low

  * depends fix

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:59:24 +0400

yabs-graphite-ceres (0.2.0.21) unstable; urgency=low

  * django versions fix

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:55:57 +0400

yabs-graphite-ceres (0.2.0.20) unstable; urgency=low

  * django versions fix

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:50:09 +0400

yabs-graphite-ceres (0.2.0.19) unstable; urgency=low

  * depends on django fixed

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:32:12 +0400

yabs-graphite-ceres (0.2.0.18) unstable; urgency=low

  * depends fixed 

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:27:56 +0400

yabs-graphite-ceres (0.2.0.17) unstable; urgency=low

  * logrotate removed, moved to yabs-graphite-ceres-config 

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 14:42:51 +0400

yabs-graphite-ceres (0.2.0.16) unstable; urgency=low

  * rollup logging switched off

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 06 Nov 2013 18:33:44 +0400

yabs-graphite-ceres (0.2.0.15) precise; urgency=low

  * read metadata only once in rollup
  * print broken metrics (broken .ceres-node json) to log 

 -- Kirill Zhalnin <kiwi@yandex-team.ru>  Fri, 15 Nov 2013 18:47:24 +0400

yabs-graphite-ceres (0.2.0.14) precise; urgency=low

  * removed bad depend for python-django-1.5 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 13 Nov 2013 14:36:01 +0400

yabs-graphite-ceres (0.2.0.13) precise; urgency=low

  * fixed shorcuts+zoom bug
  * made djandgo depend more universal

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 13 Nov 2013 14:28:23 +0400

yabs-graphite-ceres (0.2.0.12) precise; urgency=low

  * rollup stat to ceres_rollup.err 

 -- Kirill Zhalnin <kiwi@yandex-team.ru>  Thu, 07 Nov 2013 21:00:33 +0400

yabs-graphite-ceres (0.2.0.11) precise; urgency=low

  * rollup logrotate 

 -- Kirill Zhalnin <kiwi@yandex-team.ru>  Thu, 07 Nov 2013 17:02:52 +0400

yabs-graphite-ceres (0.2.0.10) precise; urgency=low

  * new rollup

 -- Kirill Zhalnin <kiwi@yandex-team.ru>  Thu, 07 Nov 2013 16:40:22 +0400

yabs-graphite-ceres (0.2.0.9) precise; urgency=low

  * fixed dashboards links

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 06 Nov 2013 17:39:13 +0400

yabs-graphite-ceres (0.2.0.8) unstable; urgency=low

  * added shortcuts for timeRange periods

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 06 Nov 2013 15:14:06 +0400

yabs-graphite-ceres (0.2.0.7) unstable; urgency=low

  * fixed series name output for diffSeries

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 30 Oct 2013 18:16:52 +0400

yabs-graphite-ceres (0.2.0.6) unstable; urgency=low

  * fixed sort order in finder

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Tue, 29 Oct 2013 13:48:43 +0400

>>>>>>> .r7382
yabs-graphite-ceres (0.2.0.5) unstable; urgency=low

  * merged to trunk 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Mon, 28 Oct 2013 15:03:24 +0400

yabs-graphite-ceres (0.2.0.4) unstable; urgency=low

  * fixed icons place 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Mon, 28 Oct 2013 14:13:48 +0400

yabs-graphite-ceres (0.2.0.3) precise; urgency=low

  * reverted bad patch

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 25 Oct 2013 13:40:26 +0400

yabs-graphite-ceres (0.2.0.2) unstable; urgency=low

  * fixed some bugs 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Thu, 24 Oct 2013 19:52:22 +0400

yabs-graphite-ceres (0.2.0.1) unstable; urgency=low

  * added correts deps 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Thu, 24 Oct 2013 19:43:06 +0400

yabs-graphite-ceres (0.2.0.0) unstable; urgency=low

  * merged with graphite trunk 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Thu, 24 Oct 2013 18:08:57 +0400

yabs-graphite-ceres (0.1.9.12) unstable; urgency=low

  [ Kulikovsky Dmitry ]
  * fixed braces bug 

  [ Vladimir Eremin ]
  * extended fix

 -- Vladimir Eremin <yottatsa@yandex-team.ru>  Tue, 22 Oct 2013 18:08:23 +0400

yabs-graphite-ceres (0.1.9.10) unstable; urgency=low

  * RTADMIN-5976 normalize

 -- Vladimir Eremin <yottatsa@yandex-team.ru>  Sun, 20 Oct 2013 20:55:33 +0400

yabs-graphite-ceres (0.1.9.9) unstable; urgency=low

  * added bsgraphite1-03h to views 

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 16 Oct 2013 17:07:47 +0400

yabs-graphite-ceres (0.1.9.8) unstable; urgency=low

  * fixed possible loop

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 20 Sep 2013 19:19:19 +0400

yabs-graphite-ceres (0.1.9.7) unstable; urgency=low

  * changed python-pytz to python-tz in deplist 

 -- Anna Sysoeva <antrue@yandex-team.ru>  Mon, 16 Sep 2013 14:12:59 +0400

yabs-graphite-ceres (0.1.9.6) unstable; urgency=low

  * json_render rethink, applied

 -- Vladimir Eremin <yottatsa@yandex-team.ru>  Tue, 10 Sep 2013 16:36:09 +0400

yabs-graphite-ceres (0.1.9.5) unstable; urgency=low

  * applied rollup patch

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Tue, 03 Sep 2013 15:06:18 +0400

yabs-graphite-ceres (0.1.9.4) unstable; urgency=low

  * fixed statLag bug

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Tue, 03 Sep 2013 14:55:17 +0400

yabs-graphite-ceres (0.1.9.3) unstable; urgency=low


webapp/graphite/settings.py
"""Copyright 2008 Orbitz WorldWide

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License."""
# Django settings for graphite project.
# DO NOT MODIFY THIS FILE DIRECTLY - use local_settings.py instead
import sys, os
from os.path import abspath, dirname, join
from warnings import warn


GRAPHITE_WEB_APP_SETTINGS_LOADED = False
WEBAPP_VERSION = '0.10.0-alpha'
DEBUG = False
JAVASCRIPT_DEBUG = False

# Filesystem layout
WEB_DIR = dirname( abspath(__file__) )
WEBAPP_DIR = dirname(WEB_DIR)
GRAPHITE_ROOT = dirname(WEBAPP_DIR)
# Initialize additional path variables
# Defaults for these are set after local_settings is imported
CONTENT_DIR = ''
CSS_DIR = ''
CONF_DIR = ''
DASHBOARD_CONF = ''
GRAPHTEMPLATES_CONF = ''
STORAGE_DIR = ''
WHITELIST_FILE = ''
INDEX_FILE = ''
LOG_DIR = ''
CERES_DIR = ''
WHISPER_DIR = ''
RRD_DIR = ''
STANDARD_DIRS = []

CLUSTER_SERVERS = []

sys.path.insert(0, WEBAPP_DIR)

# Cluster settings
CLUSTER_SERVERS = []
REMOTE_FIND_TIMEOUT = 3.0
REMOTE_FETCH_TIMEOUT = 6.0
REMOTE_RETRY_DELAY = 60.0
REMOTE_READER_CACHE_SIZE_LIMIT = 1000
CARBONLINK_HOSTS = ["127.0.0.1:7002"]
CARBONLINK_TIMEOUT = 1.0
CARBONLINK_HASHING_KEYFUNC = None
CARBONLINK_RETRY_DELAY = 15
REPLICATION_FACTOR = 1
MEMCACHE_HOSTS = []
FIND_CACHE_DURATION = 300
FIND_TOLERANCE = 2 * FIND_CACHE_DURATION
DEFAULT_CACHE_DURATION = 60 #metric data and graphs are cached for one minute by default
LOG_CACHE_PERFORMANCE = False

MAX_FETCH_RETRIES = 2

#Remote rendering settings
REMOTE_RENDERING = False #if True, rendering is delegated to RENDERING_HOSTS
RENDERING_HOSTS = []
REMOTE_RENDER_CONNECT_TIMEOUT = 1.0
LOG_RENDERING_PERFORMANCE = False

#Miscellaneous settings
SMTP_SERVER = "localhost"
DOCUMENTATION_URL = "http://graphite.readthedocs.org/"
ALLOW_ANONYMOUS_CLI = True
LOG_METRIC_ACCESS = False
LEGEND_MAX_ITEMS = 10
RRD_CF = 'AVERAGE'
STORAGE_FINDERS = (
    'graphite.finders.CeresFinder',
    'graphite.finders.StandardFinder',
)

#Authentication settings
USE_LDAP_AUTH = False
LDAP_SERVER = "" # "ldapserver.mydomain.com"
LDAP_PORT = 389
LDAP_USE_TLS = False
LDAP_SEARCH_BASE = "" # "OU=users,DC=mydomain,DC=com"
LDAP_BASE_USER = "" # "CN=some_readonly_account,DC=mydomain,DC=com"
LDAP_BASE_PASS = "" # "my_password"
LDAP_USER_QUERY = "" # "(username=%s)"  For Active Directory use "(sAMAccountName=%s)"
LDAP_URI = None

#Set this to True to delegate authentication to the web server
USE_REMOTE_USER_AUTHENTICATION = False

# Django 1.5 requires this so we set a default but warn the user
SECRET_KEY = 'UNSAFE_DEFAULT'

# Django 1.5 requires this to be set. Here we default to prior behavior and allow all
ALLOWED_HOSTS = [ '*' ]

# Override to link a different URL for login (e.g. for django_openid_auth)
LOGIN_URL = '/account/login'

# Set to True to require authentication to save or delete dashboards
DASHBOARD_REQUIRE_AUTHENTICATION = False
# Require Django change/delete permissions to save or delete dashboards.
# NOTE: Requires DASHBOARD_REQUIRE_AUTHENTICATION to be set
DASHBOARD_REQUIRE_PERMISSIONS = False
# Name of a group to which the user must belong to save or delete dashboards.  Alternative to
# DASHBOARD_REQUIRE_PERMISSIONS, particularly useful when using only LDAP (without Admin app)
# NOTE: Requires DASHBOARD_REQUIRE_AUTHENTICATION to be set
DASHBOARD_REQUIRE_EDIT_GROUP = None

DATABASES = {
  'default': {
    'NAME': '/opt/graphite/storage/graphite.db',
    'ENGINE': 'django.db.backends.sqlite3',
    'USER': '',
    'PASSWORD': '',
    'HOST': '',
    'PORT': '',
  },
}

# If using rrdcached, set to the address or socket of the daemon
FLUSHRRDCACHED = ''

## Load our local_settings
try:
  from graphite.local_settings import *
except ImportError:
  print >> sys.stderr, "Could not import graphite.local_settings, using defaults!"

## Load Django settings if they werent picked up in local_settings
if not GRAPHITE_WEB_APP_SETTINGS_LOADED:
  from graphite.app_settings import *

## Set config dependent on flags set in local_settings
# Path configuration
if not CONTENT_DIR:
  CONTENT_DIR = join(WEBAPP_DIR, 'content')
if not CSS_DIR:
  CSS_DIR = join(CONTENT_DIR, 'css')

if not CONF_DIR:
  CONF_DIR = os.environ.get('GRAPHITE_CONF_DIR', join(GRAPHITE_ROOT, 'conf'))
if not DASHBOARD_CONF:
  DASHBOARD_CONF = join(CONF_DIR, 'dashboard.conf')
if not GRAPHTEMPLATES_CONF:
  GRAPHTEMPLATES_CONF = join(CONF_DIR, 'graphTemplates.conf')

if not STORAGE_DIR:
  STORAGE_DIR = os.environ.get('GRAPHITE_STORAGE_DIR', join(GRAPHITE_ROOT, 'storage'))
if not WHITELIST_FILE:
  WHITELIST_FILE = join(STORAGE_DIR, 'lists', 'whitelist')
if not INDEX_FILE:
  INDEX_FILE = join(STORAGE_DIR, 'index')
if not LOG_DIR:
  LOG_DIR = join(STORAGE_DIR, 'log', 'webapp')
if not WHISPER_DIR:
  WHISPER_DIR = join(STORAGE_DIR, 'whisper/')
if not CERES_DIR:
  CERES_DIR = join(STORAGE_DIR, 'ceres/')
if not RRD_DIR:
  RRD_DIR = join(STORAGE_DIR, 'rrd/')
if not STANDARD_DIRS:
  try:
    import whisper
    if os.path.exists(WHISPER_DIR):
      STANDARD_DIRS.append(WHISPER_DIR)
  except ImportError:
    print >> sys.stderr, "WARNING: whisper module could not be loaded, whisper support disabled"
  try:
    import rrdtool
    if os.path.exists(RRD_DIR):
      STANDARD_DIRS.append(RRD_DIR)
  except ImportError:
    pass

# Default sqlite db file
# This is set here so that a user-set STORAGE_DIR is available
if 'sqlite3' in DATABASES.get('default',{}).get('ENGINE','') \
    and not DATABASES.get('default',{}).get('NAME'):
  DATABASES['default']['NAME'] = join(STORAGE_DIR, 'graphite.db')

# Caching shortcuts
if MEMCACHE_HOSTS:
  CACHE_BACKEND = 'memcached://' + ';'.join(MEMCACHE_HOSTS) + ('/?timeout=%d' % DEFAULT_CACHE_DURATION)

# Authentication shortcuts
if USE_LDAP_AUTH and LDAP_URI is None:
  LDAP_URI = "ldap://%s:%d/" % (LDAP_SERVER, LDAP_PORT)

if USE_REMOTE_USER_AUTHENTICATION:
  MIDDLEWARE_CLASSES += ('django.contrib.auth.middleware.RemoteUserMiddleware',)
  AUTHENTICATION_BACKENDS.insert(0,'django.contrib.auth.backends.RemoteUserBackend')

if USE_LDAP_AUTH:
  AUTHENTICATION_BACKENDS.insert(0,'graphite.account.ldapBackend.LDAPBackend')

if SECRET_KEY == 'UNSAFE_DEFAULT':
  warn('SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security')
